<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [963]:
import pandas as pd
import psycopg2

In [964]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [965]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [966]:
query_3_1 = f'''
-- SQL-запрос для подсчета количества вакансий в таблице vacancies
SELECT COUNT(*) -- подсчитывает общее количество строк в таблице
FROM VACANCIES -- указывает таблицу vacancies как источник данных для запроса
'''

In [967]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_3_1, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\2762814221.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [968]:
query_3_2 = f'''
-- SQL-запрос для подсчета количества работодателей в таблице employers
SELECT COUNT(*) -- подсчитываем общее количество строк в таблице
FROM EMPLOYERS -- указываем таблицу employers как источник данных для запроса
'''

In [969]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_3_2, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\961071152.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [970]:
query_3_3 = f'''
-- SQL-запрос для подсчета количества регионов в таблице areas
SELECT COUNT(*) -- подсчитываем общее количество строк в таблице
FROM AREAS -- указываем таблицу areas как источник данных для запроса
'''

In [971]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_3_3, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\3118071541.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [972]:
query_3_4 = f'''
-- SQL-запрос для подсчета количества сфер деятельности в таблице industries
SELECT COUNT(*) -- подсчитываем общее количество строк в таблице
FROM INDUSTRIES -- указываем таблицу industries как источник данных для запроса
'''

In [973]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_3_4, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\2301577313.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    294

***

**Выводы предварительного анализа данных:**

- Рынок труда: В базе данных содержится 49,197 вакансий, что указывает на большое количество возможностей для соискателей. Это число может отражать активность рынка труда в определённый период времени или в определённой географической области.  
- Работодатели: С 23,501 работодателем в базе данных, видно, что множество компаний ищут новых сотрудников. Это разнообразие может способствовать конкуренции среди работодателей за привлечение талантов.  
- Географическое Распределение: Существует 1,362 региона, что может указывать на широкий географический охват базы данных. Это также может быть полезно для анализа рынка труда по различным регионам и выявления трендов или потребностей в определённых местах.  
- Сферы Деятельности: Наличие 294 сфер деятельности показывает многообразие отраслей, представленных в базе данных. Это может быть полезно для анализа спроса на рабочую силу в различных отраслях и планирования карьерного роста.  
Эти данные могут быть полезны для разработки стратегий поиска работы, планирования карьеры, а также для компаний, стремящихся понять рынок труда и оптимизировать свои процессы найма. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [974]:
query_4_1 = f'''
-- Запрос для определения пяти регионов с наибольшим количеством вакансий
SELECT 
    area.name AS "Region Name", -- 'Region Name' - название региона
    COUNT(job.id) AS "Vacancies Count" -- 'Vacancies Count' - количество вакансий в регионе
FROM AREAS area -- 'AREAS' - таблица, содержащая информацию о регионах
INNER JOIN VACANCIES job ON area.id = job.area_id -- Соединение с таблицей 'vacancies' для связи вакансий с регионами
GROUP BY area.id -- Группировка результатов по уникальному идентификатору региона
ORDER BY "Vacancies Count" DESC -- Сортировка результатов по убыванию количества вакансий
LIMIT 5 -- Ограничение вывода до пяти регионов с наибольшим количеством вакансий
'''

In [975]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_4_1, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\509453402.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Region Name  Vacancies Count
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [976]:
query_4_2 = f'''
-- Запрос для подсчета вакансий с указанными данными о зарплате
SELECT COUNT(*) AS "Total Vacancies with Salary Info"
FROM VACANCIES -- Таблица, содержащая информацию о вакансиях
WHERE NOT (salary_from IS NULL AND salary_to IS NULL) -- Условие, что хотя бы одно поле зарплаты заполнено
'''
#В SQL логика оператора AND в условии WHERE NOT используется для выбора записей, 
#где оба условия в скобках должны быть истинными, чтобы вся конструкция стала ложной, и наоборот.
#Это называется законом Де Моргана в булевой алгебре, который гласит, 
#что отрицание конъюнкции NOT (... AND ...) равно дизъюнкции отрицаний (... NOT OR ... NOT), и наоборот.

In [977]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_4_2, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\3612891776.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Total Vacancies with Salary Info
0                             24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [978]:
# текст запроса
query_4_3 = f'''
-- Определение средних значений зарплатных предложений
SELECT 
    ROUND(AVG(salary_from), 0) AS "Average_Min_Salary", -- Округление средней минимальной зарплаты до ближайшего целого числа
    ROUND(AVG(salary_to), 0) AS "Average_Max_Salary"    -- Округление средней максимальной зарплаты до ближайшего целого числа
FROM VACANCIES -- Работа с данными из таблицы вакансий
'''

In [979]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_4_3, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\2374954388.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Average_Min_Salary  Average_Max_Salary
0             71065.0            110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [980]:
query_4_4 = f'''
-- Запрос для определения количества вакансий по каждой комбинации графика работы и типа трудоустрйоства
SELECT 
    v.schedule AS "Work Schedule", -- 'Work Schedule' - тип графика работы
    v.employment AS "Employment Type", -- 'Employment Type' - форма занятости
    COUNT(v.id) AS "Vacancies Count" -- 'Vacancies Count' - подсчет количества вакансий для каждой комбинации
FROM 
    VACANCIES v -- 'VACANCIES' - таблица с данными о вакансиях
GROUP BY 
    "Work Schedule", "Employment Type" -- Группировка по типу графика и форме занятости
ORDER BY 
    "Vacancies Count" DESC -- Сортировка по убыванию количества вакансий
OFFSET 
    1 -- Пропуск первой строки в списке
LIMIT 
    1 -- Вывод только одной строки
'''

In [981]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_4_4, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\481693114.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Work Schedule   Employment Type  Vacancies Count
0  Удаленная работа  Полная занятость             7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [982]:
# текст запроса
query_4_5 = f'''
-- Получаем требуемый опыт работы и соответствующее количество вакансий
SELECT 
    experience AS "Experience", -- 'Требуемый опыт работы' для каждой вакансии
    COUNT(*) AS "VacanciesCount"        -- 'Количество вакансий' с таким опытом

-- Делаем выборку из таблицы с вакансиями
FROM 
    VACANCIES                           -- Таблица 'vacancies' содержит данные о вакансиях

-- Группировка данных по опыту работы
GROUP BY 
    experience                          -- Группировка по 'experience' для агрегации данных

-- Упорядочивание результатов по количеству вакансий в порядке увеличения
ORDER BY 
    "VacanciesCount" ASC;               -- Сортировка по возрастанию 'VacanciesCount
'''

In [983]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5 # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\2298421073.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Experience  VacanciesCount
0         Более 6 лет            1337
1           Нет опыта            7197
2       От 3 до 6 лет           14511
3  От 1 года до 3 лет           26152

***

**Выводы детального анализа вакансий**: 

1. Региональный распределение вакансий:
Москва лидирует с 5333 вакансиями, что может быть связано с её статусом экономического и политического центра России.  
Санкт-Петербург следует за Москвой с 2851 вакансиями, подтверждая его значимость как крупного культурного и промышленного города.  
Минск, Новосибирск и Алматы также показывают высокое количество вакансий, что может отражать растущую экономическую активность в этих регионах.  

2. Вакансии с информацией о зарплате:  
Из 24073 вакансий, информация о зарплате указана хотя бы в одном поле, что может свидетельствовать о тенденции работодателей быть более открытыми в вопросах оплаты труда.  

3. Средние значения зарплат:  
Средняя минимальная зарплата составляет 71065 рублей, а средняя максимальная — 110537 рублей. Это различие может указывать на широкий диапазон возможностей для соискателей с разным уровнем квалификации и опыта. 

4. Типы рабочего графика и трудоустройства:  
Наибольшее количество вакансий (7802) предлагается для удаленной работы с полной занятостью, что может отражать текущие тенденции рынка труда и желание работников иметь гибкий график.  

5. Требуемый опыт работы:  
Большинство вакансий (26152) требуют от 1 года до 3 лет опыта, что может свидетельствовать о потребности рынка в молодых специалистах, готовых к обучению и развитию.  
Гипотезы и предложения:  

Работодатели могут увеличить привлекательность вакансий, предлагая более конкурентоспособные зарплаты и гибкие условия труда.  
Увеличение количества вакансий с требованием “нет опыта” может способствовать снижению безработицы среди молодежи.  
Рынок труда демонстрирует рост в крупных городах с развитой инфраструктурой и экономикой. Тенденция к удаленной работе и гибким условиям труда становится все более заметной, а спрос на молодых специалистов подчеркивает важность инвестиций в образование и профессиональное развитие.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [984]:
query_5_1 = f'''
-- Создаем рейтинг работодателей по количеству предложений работы
WITH EmployerRanking AS (
    SELECT 
        e.name AS "EmployerName", -- 'EmployerName' - название фирмы
        COUNT(v.id) AS "TotalVacancies", -- 'TotalVacancies' - всего вакансий
        DENSE_RANK() OVER (ORDER BY COUNT(v.id) DESC) AS "Position" -- 'Position' - место по количеству вакансий
    FROM VACANCIES v -- 'VACANCIES' - таблица с объявлениями о работе
    JOIN EMPLOYERS e ON e.id = v.employer_id -- соединяем с таблицей работодателей
    GROUP BY e.name -- группируем по названию компании
)
-- Получаем информацию о работодателях, занимающих первое и пятое места
SELECT "EmployerName", "TotalVacancies"
FROM EmployerRanking
WHERE "Position" IN (1, 5) -- выбираем первое и пятое места
'''

In [985]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_5_1, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\1914783865.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



EmployerName  TotalVacancies
0         Яндекс            1933
1  Газпром нефть             331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [986]:
query_5_2 = f'''
-- Определяем статистику по работодателям и вакансиям для каждого региона
SELECT 
    a.name AS "RegionName", -- 'RegionName' - идентификатор региона
    COUNT(DISTINCT e.id) AS "EmployersCount", -- 'EmployersCount' - число работодателей в регионе
    COUNT(DISTINCT v.id) AS "VacanciesCount" -- 'VacanciesCount' - число вакансий в регионе
FROM AREAS a
LEFT JOIN VACANCIES v ON v.area_id = a.id -- Связываем регионы с вакансиями
LEFT JOIN EMPLOYERS e ON e.area = a.id -- Связываем регионы с работодателями
WHERE v.id is NULL -- Фильтрация регионов, где нет вакансий
GROUP BY a.name -- Группировка данных по названию региона
ORDER BY "EmployersCount" DESC -- Сортируем данные сначала по числу работодателей
LIMIT 1 -- Выводим только регион с наибольшим числом работодателей, где нет вакансий
'''

In [987]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_5_2, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\204662484.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



RegionName  EmployersCount  VacanciesCount
0     Россия             410               0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [988]:
query_5_3 = f'''
-- Вычисляем распределение работодателей по регионам
SELECT 
    employer_id AS "EmployerID", -- 'EmployerID' - уникальный номер работодателя
    COUNT(DISTINCT area_id) AS "RegionsCount" -- 'RegionsCount' - число регионов с вакансиями
FROM 
    VACANCIES -- 'VACANCIES' - таблица, содержащая данные о вакансиях
GROUP BY 
    "EmployerID" -- Группировка по идентификатору работодателя
ORDER BY 
    "RegionsCount" DESC -- Сортировка по количеству регионов в убывающем порядке
LIMIT 
    1 -- Выводим только верхнюю запись
'''

In [989]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_5_3, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\2536702449.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



EmployerID  RegionsCount
0        1740           181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [990]:
query_5_4 = f'''
-- Рассчитываем число компаний без указания сферы деятельности
SELECT 
    COUNT(*) AS "EmployersWithoutIndustry" -- 'EmployersWithoutIndustry' - количество работодателей без сферы деятельности
FROM 
    EMPLOYERS e -- 'EMPLOYERS' - таблица, содержащая информацию о работодателях
LEFT JOIN 
    EMPLOYERS_INDUSTRIES ei ON e.id = ei.employer_id -- связываем данные о работодателях с их отраслями
WHERE 
    ei.industry_id IS NULL -- фильтруем работодателей, у которых не указана отрасль
'''

In [991]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_5_4, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\4162815337.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



EmployersWithoutIndustry
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [992]:
query_5_5 = f'''
-- Определяем компании с четырьмя сферами деятельности и их позицию в алфавитном списке
SELECT 
    e.name AS "CompanyName" -- 'CompanyName' - название компании
FROM 
    EMPLOYERS e
JOIN (
    -- Фильтруем компании по количеству сфер деятельности
    SELECT 
        employer_id
    FROM 
        EMPLOYERS_INDUSTRIES
    GROUP BY 
        employer_id
    HAVING 
        COUNT(industry_id) = 4 -- Выбираем компании с четырьмя отраслями
) AS FilteredEmployers ON e.id = FilteredEmployers.employer_id
ORDER BY 
    "CompanyName" -- Сортировка компаний по алфавиту
LIMIT 
    1 OFFSET 2 -- Пропускаем две компании для получения третьей в списке
'''

In [993]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_5_5, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\3502007031.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



CompanyName
0        2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [994]:
query_5_6 = f'''
-- Определяем количество компаний, работающих в области IT
SELECT 
    COUNT(*) AS "SoftwareDevEmployers" -- 'SoftwareDevEmployers' - число работодателей в IT
FROM 
    EMPLOYERS_INDUSTRIES ei -- 'EMPLOYERS_INDUSTRIES' - таблица отношений работодателей и индустрий
WHERE 
    EXISTS ( -- Проверяем наличие записей
        SELECT 1
        FROM INDUSTRIES i -- 'INDUSTRIES' - таблица индустрий
        WHERE 
            i.id = ei.industry_id -- Сопоставляем индустрии с работодателями
            AND i.name = 'Разработка программного обеспечения' -- Фильтруем по названию индустрии
    )
'''

In [995]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_5_6, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\1993108486.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



SoftwareDevEmployers
0                  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [996]:
# Импортируем необходимые библиотеки
import requests
from bs4 import BeautifulSoup

# Указываем URL страницы Википедии с городами-миллионерами России
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
# Отправляем GET-запрос на URL
response = requests.get(url)
# Парсим полученную страницу с помощью BeautifulSoup
page = BeautifulSoup(response.text, 'html.parser')
# Ищем таблицу с классом "standard" и преобразуем её в строку
cities_data = str(page.find('table', class_="standard"))
# Разделяем строку по атрибуту 'title="' и берем все элементы начиная с третьего
cities_data = cities_data.split('title="')[3:]

# Создаем пустой список для хранения названий городов
cities_list = []
# Проходимся по каждому элементу в списке cities_data
for city_info in cities_data:
    # Добавляем в список cities_list название города, разделив строку по '">'
    cities_list.append(city_info.split('">')[0])
# Преобразуем список городов в кортеж
cities_tuple = tuple(cities_list)


In [997]:
query_5_7 = f'''
-- Создаем временную таблицу с ранжированием работодателей
WITH RankedEmployers AS (
    -- Выбираем имя региона и количество вакансий
    SELECT a.name AS AreaName,
           COUNT(*) AS VacanciesCount
    -- Присоединяем таблицы работодателей и вакансий
    FROM employers e
    JOIN vacancies v ON e.id = v.employer_id
    -- Присоединяем таблицу регионов
    JOIN areas a ON a.id = v.area_id
    -- Условие фильтрации по имени работодателя и вхождению региона в кортеж городов
    WHERE e.name = 'Яндекс'
          AND a.name IN {cities_tuple}
    -- Группируем результаты по идентификатору региона
    GROUP BY a.id
)
-- Выбираем имя региона и количество вакансий из временной таблицы
SELECT AreaName, 
       VacanciesCount
FROM RankedEmployers
-- Добавляем к результату общее количество вакансий
UNION ALL
-- Выбираем лейбл 'Total' и сумму количества вакансий
SELECT 'Total' AS TotalLabel, 
       SUM(VacanciesCount) AS TotalCount
FROM RankedEmployers
'''

In [998]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_5_7, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\1272580378.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



areaname  vacanciescount
0            Москва            54.0
1   Санкт-Петербург            42.0
2      Екатеринбург            39.0
3       Новосибирск            35.0
4         Волгоград            24.0
5           Воронеж            32.0
6         Краснодар            30.0
7        Красноярск            23.0
8   Нижний Новгород            36.0
9              Омск            21.0
10            Пермь            25.0
11   Ростов-на-Дону            25.0
12           Самара            26.0
13           Казань            25.0
14              Уфа            26.0
15        Челябинск            22.0
16            Total           485.0

***

**Выводы по анализу работодателей**:  

1. Лидеры по количеству вакансий:  
Результаты показывают, что Яндекс занимает первое место с 1933 вакансиями, а Газпром нефть — пятое с 331 вакансиями. Это может указывать на то, что Яндекс активно расширяется и имеет большое количество проектов, требующих новых сотрудников. Газпром нефть, хотя и занимает более низкую позицию, тем не менее, является крупным работодателем в своей отрасли.
2. Работодатели без вакансий:
Самый большой интерес представляет регион Россия с 410 работодателями и нулевым количеством вакансий. Это может быть связано с тем, что данные были собраны не из всех возможных источников или вакансии были закрыты до момента сбора данных.
3. Распределение работодателей по регионам:
Работодатель с идентификатором 1740 представлен в 181 регионе, что указывает на его широкий охват и возможно национальное значение.
4. Работодатели без сферы деятельности:
Наличие 8419 работодателей без указания сферы деятельности может указывать на недостатки в системе сбора данных или на то, что многие компании не предоставляют полную информацию о своей деятельности.
5. Компании с четырьмя сферами деятельности:
Компания 2ГИС находится на третьем месте в алфавитном списке среди компаний с четырьмя сферами деятельности. Это может говорить о разнообразии их бизнеса и о том, что они занимаются различными видами деятельности.
6. Работодатели в сфере IT:
3553 работодателя указали сферу деятельности как "Разработка программного обеспечения", что подчеркивает значимость IT-отрасли и высокий спрос на специалистов в этой области.
7. Вакансии “Яндекс” в городах-миллионниках:
Анализ вакансий Яндекса в городах-миллионниках показывает, что большинство вакансий сосредоточено в Москве (54) и Санкт-Петербурге (42), что логично, учитывая их статус крупнейших экономических центров России. Общее количество вакансий составляет 485, что может свидетельствовать о сильной позиции компании на рынке труда.

Яндекс и другие крупные компании могут рассмотреть возможность расширения присутствия в регионах с меньшим количеством вакансий для привлечения талантов.
Работодателям следует улучшить качество данных о своих вакансиях, чтобы обеспечить более точный анализ и привлечение кандидатов.
IT-отрасль продолжает оставаться важной для экономики, и компании должны инвестировать в развитие технологических навыков сотрудников.

Анализ данных о работодателях и вакансиях показывает, что крупные компании, такие как Яндекс и Газпром нефть, доминируют на рынке труда. Регионы с большим количеством работодателей, но без вакансий, могут указывать на потенциальные возможности для роста занятости. IT-сектор остается ключевым, и есть потребность в улучшении качества данных для более эффективного соединения работодателей и соискателей.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [999]:
query_6_1 = f'''
-- Подсчет количества вакансий, связанных с обработкой данных
SELECT 
    COUNT(*) AS "DataRelatedVacancies" -- 'DataRelatedVacancies' - общее количество вакансий, связанных с данными
FROM 
    Vacancies -- 'Vacancies' - таблица, содержащая информацию о вакансиях
WHERE 
    UPPER(name) LIKE '%DATA%' -- Поиск вакансий, содержащих слово 'DATA' в названии, без учета регистра
    OR 
    UPPER(name) LIKE '%ДАНН%' -- Поиск вакансий, содержащих слово 'ДАНН' в названии, также без учета регистра
'''

In [1000]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_6_1, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\2857538695.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



DataRelatedVacancies
0                  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [1001]:
query_6_2 = f'''
-- Выборка количества вакансий для начинающих специалистов в области данных
SELECT 
    COUNT(*) AS "EntryLevelVacancies" -- 'EntryLevelVacancies' - количество вакансий для начинающих
FROM 
    Vacancies -- 'Vacancies' - таблица вакансий
WHERE 
    (
        LOWER(name) LIKE '%data scientist%' OR -- Ищем вакансии для 'data scientist'
        LOWER(name) LIKE '%data science%' OR -- Ищем вакансии в области 'data science'
        LOWER(name) LIKE '%исследователь данных%' OR -- Ищем вакансии 'исследователя данных'
        (
            LOWER(name) LIKE '%ml%' AND -- Ищем вакансии по 'ML', исключая те, что связаны с HTML
            LOWER(name) NOT LIKE '%html%'
        ) OR
        LOWER(name) LIKE '%machine learning%' OR -- Ищем вакансии 'machine learning'
        LOWER(name) LIKE '%машинн%обучен%' -- Ищем вакансии по 'машинному обучению'
    )
    AND
    (
        LOWER(name) LIKE '%junior%' OR -- Ищем вакансии для 'junior' специалистов
        LOWER(experience) LIKE '%нет опыта%' OR -- Ищем вакансии без требования опыта
        LOWER(employment) LIKE '%стажировка%' -- Ищем стажировки
    )

''' 

In [1002]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_6_2, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\1939716864.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



EntryLevelVacancies
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [1003]:
query_6_3 = f'''
-- Выборка количества вакансий для специалистов в области Data Science с упором на SQL и Postgres
SELECT 
    COUNT(*) AS "VacanciesWithSQL" -- 'VacanciesWithSQL' - количество вакансий, где требуется SQL или Postgres
FROM 
    Vacancies -- 'Vacancies' - таблица, содержащая информацию о вакансиях
WHERE 
    (
        LOWER(name) LIKE '%data scientist%' OR -- Ищем вакансии для 'Data Scientist'
        LOWER(name) LIKE '%data science%' OR -- Ищем вакансии в области 'Data Science'
        LOWER(name) LIKE '%исследователь данных%' OR -- Ищем вакансии 'Исследователя данных'
        LOWER(name) LIKE '%machine learning%' OR -- Ищем вакансии 'Machine Learning'
        LOWER(name) LIKE '%машинн%обучен%' OR -- Ищем вакансии 'Машинного обучения'
        (LOWER(name) LIKE '%ml%' AND NOT LOWER(name) LIKE '%html%') -- Ищем вакансии по 'ML', исключая 'HTML'
    ) AND (
        LOWER(key_skills) LIKE '%sql%' OR -- Ключевые навыки включают 'SQL'
        LOWER(key_skills) LIKE '%postgres%' -- Или 'Postgres'
    )
'''

In [1004]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_6_3, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\1711649379.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



VacanciesWithSQL
0               229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [1005]:
# текст запроса
query_6_4 = f'''
-- Определение количества вакансий для специалистов Data Science с требованием знания Python
SELECT 
    COUNT(*) AS "PythonVacancies" -- 'PythonVacancies' - количество вакансий, где требуется Python
FROM 
    Vacancies -- 'Vacancies' - таблица, содержащая информацию о вакансиях
WHERE 
    (
        LOWER(name) LIKE '%data scientist%' OR -- Ищем вакансии для 'Data Scientist'
        LOWER(name) LIKE '%data science%' OR -- Ищем вакансии в области 'Data Science'
        LOWER(name) LIKE '%исследователь данных%' OR -- Ищем вакансии 'Исследователя данных'
        LOWER(name) LIKE '%machine learning%' OR -- Ищем вакансии 'Machine Learning'
        LOWER(name) LIKE '%машинн%обучен%' OR -- Ищем вакансии 'Машинного обучения'
        (LOWER(name) LIKE '%ml%' AND NOT LOWER(name) LIKE '%html%') -- Ищем вакансии по 'ML', исключая 'HTML'
    ) AND (
        LOWER(key_skills) LIKE '%python%' -- Ключевые навыки включают 'Python'
    )
'''


In [1006]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_6_4, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\1129006649.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



PythonVacancies
0              357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [1007]:
query_6_5 = f'''
SELECT 
    -- Округляем среднее значение до двух знаков после запятой
    ROUND(AVG(
        -- Вычисляем длину строки key_skills
        LENGTH(key_skills) - 
        -- Вычитаем длину строки key_skills без табуляции (CHR(9))
        LENGTH(REPLACE(key_skills, CHR(9), '')) + 1
    ), 2)
FROM 
    VACANCIES
WHERE 
    -- Фильтруем вакансии по названию, связанным с Data Science
    (name ILIKE '%data scientist%' OR 
     name ILIKE '%data science%' OR 
     name ILIKE '%исследователь данных%' OR 
     -- Исключаем вакансии, связанные с HTML, но включаем ML
     (name LIKE '%ML%' AND 
      name NOT LIKE '%HTML%') OR
     name ILIKE '%machine learning%' OR 
     name ILIKE '%машинн%обучен%')
'''

In [1008]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df = pd.read_sql_query(query_6_5, connection)
df # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\4097510479.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [1009]:
query_6_6 = f'''
SELECT 
    -- Выбираем столбец опыта работы для группировки результатов
    experience AS experience, 
    -- Рассчитываем среднюю зарплату, округляя до целого числа
    ROUND(AVG(
        -- Используем COALESCE для выбора первого ненулевого значения
        COALESCE(
            -- Среднее значение между salary_from и salary_to, если оба доступны
            (salary_from + salary_to) / 2, 
            salary_from, 
            salary_to
        )
    ), 0) AS salary_ds 
FROM 
    VACANCIES -- Из таблицы вакансий
WHERE 
    -- Фильтруем вакансии по названию, связанным с Data Science
    (name ILIKE '%data scientist%' OR 
     name ILIKE '%data science%' OR 
     name ILIKE '%исследователь данных%' OR 
     -- Исключаем вакансии, связанные с HTML, но включаем ML
     (name LIKE '%ML%' AND 
      name NOT LIKE '%HTML%') OR
     name ILIKE '%machine learning%' OR 
     name ILIKE '%машинн%обучен%') AND  
    -- Условие, что зарплата должна быть указана
    (salary_from IS NOT NULL OR salary_to IS NOT NULL) 
GROUP BY 
    -- Группируем результаты по опыту работы
    experience; 
'''

In [1010]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df66 = pd.read_sql_query(query_6_6, connection)
df66 # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\80771059.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  salary_ds
0           Нет опыта    74643.0
1  От 1 года до 3 лет   139675.0
2       От 3 до 6 лет   243115.0

In [1011]:
avg_salary = int(df66[df66['experience'] == 'От 3 до 6 лет']['salary_ds'].iloc[0])
display(f'Средняя зарплата DS с опытом работы от 3 до 6 лет: {avg_salary}')

'Средняя зарплата DS с опытом работы от 3 до 6 лет: 243115'

***

In [1012]:
import plotly.express as px

fig = px.line(df66, x='experience', y='salary_ds', title='График зависимости средней зарплаты DS от опыта работы')
fig.update_layout(xaxis_title='Опыт работы, лет', yaxis_title='Средняя зарплата DS, руб', grid=dict(rows=1, columns=2))
fig.show()

**7. Дополнительное задание:**  
Выведим список из пяти работодателей с наивысшей средней зарплатой по их вакансиям, учитывая только тех работодателей, которые размещают вакансии более чем в одном регионе. В расчете средней зарплаты исключены вакансии без указания зарплаты. Результат отсортируем по убыванию средней зарплаты.

In [1013]:
query_6_7 = f'''
SELECT
    E.name AS EmployerName, -- выбираем название работодателя
    AVG(COALESCE(V.salary_from, V.salary_to, (V.salary_from + V.salary_to) / 2)) AS AverageSalary, -- считаем среднюю зарплату, исключая NULL значения
    STRING_AGG(DISTINCT V.experience, ', ') AS Experience, -- агрегируем уникальные значения опыта работы
    STRING_AGG(DISTINCT V.schedule, ', ') AS Schedule -- агрегируем уникальные значения графика работы
FROM
    EMPLOYERS E -- указываем таблицу работодателей как источник данных
JOIN
    VACANCIES V ON E.id = V.employer_id -- соединяем таблицу вакансий с таблицей работодателей по идентификатору работодателя
JOIN
    AREAS A ON V.area_id = A.id -- соединяем таблицу вакансий с таблицей регионов по идентификатору региона
WHERE
    V.salary_from IS NOT NULL OR V.salary_to IS NOT NULL -- фильтруем вакансии, где указана хотя бы одна из зарплат
GROUP BY
    E.name -- группируем результаты по названию работодателя
HAVING
    COUNT(DISTINCT A.id) > 1 -- включаем только тех работодателей, которые имеют вакансии более чем в одном регионе
ORDER BY
    AverageSalary DESC -- сортируем результаты по убыванию средней зарплаты
LIMIT 5; -- ограничиваем вывод первыми пятью строками
'''

In [1014]:
# выполняем SQL-запрос и сохраняем результат в DataFrame 'df'
df67 = pd.read_sql_query(query_6_7, connection)
df67 # отображаем содержимое DataFrame 'df'

C:\Users\Администратор\AppData\Local\Temp\ipykernel_14920\3194453808.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employername  averagesalary  \
0              Август, Фирма       500000.0   
1                   HR Prime       453517.6   
2                      Aston       375000.0   
3  Anchous Drive Recruitment       350691.0   
4                  oneFactor       350000.0   

                          experience                       schedule  
0                        Более 6 лет                    Полный день  
1  От 1 года до 3 лет, От 3 до 6 лет  Полный день, Удаленная работа  
2         Более 6 лет, От 3 до 6 лет                    Полный день  
3         Более 6 лет, От 3 до 6 лет               Удаленная работа  
4           Нет опыта, От 3 до 6 лет  Полный день, Удаленная работа

Существует значительное разнообразие в зарплатных предложениях и требованиях к опыту работы среди различных работодателей. Работодатели, предлагающие более высокие зарплаты, как правило, ищут кандидатов с более значительным опытом работы, что может указывать на более сложные и ответственные должности в их компаниях. Различия в средней зарплате могут быть обусловлены разными уровнями ответственности, квалификации и требуемого опыта для вакансий. Наличие вакансий с разными требованиями к опыту работы может указывать на широкий спектр должностей, открытых в компаниях. Вариативность графика работы отражает современные тенденции гибкости в трудоустройстве.

**Выводы по предметному анализу**:  

1. Общее количество вакансий, связанных с данными: В базе данных насчитывается 1771 вакансия, которая имеет отношение к данным. Это говорит о значительном спросе на специалистов, работающих с данными, что подтверждает важность навыков анализа данных на современном рынке труда.
2. Вакансии для начинающих дата-сайентистов: Количество вакансий для начинающих специалистов составляет 51. Это может указывать на то, что вход в профессию дата-сайентиста относительно доступен для новичков, и есть возможности для стажировок и позиций без требования опыта работы.
3. Вакансии с требованием знания SQL или Postgres: Наличие 229 вакансий, где в качестве ключевого навыка указан SQL или Postgres, подчеркивает значимость этих технологий в работе с данными. SQL остается фундаментальным инструментом для работы с базами данных, а Postgres — популярной системой управления базами данных.
4. Популярность Python среди работодателей: С 357 вакансиями, где требуется знание Python, можно сделать вывод, что Python является одним из самых востребованных языков программирования в сфере DS. Это связано с его гибкостью, мощными библиотеками для анализа данных и машинного обучения, а также с широким сообществом разработчиков.
5. Среднее количество ключевых навыков в вакансиях: В среднем в вакансиях для DS указывают 6.41 ключевых навыков. Это отражает мультидисциплинарный характер профессии, где от кандидатов ожидается знание различных инструментов и методологий.
6. Зарплата в зависимости от опыта работы:
Без опыта: Средняя зарплата составляет 74 643 рублей.
От 1 года до 3 лет: Средняя зарплата увеличивается до 139 675 рублей.
От 3 до 6 лет: Средняя зарплата достигает 243 115 рублей.
Эти данные показывают, что с увеличением опыта работы зарплата дата-сайентистов значительно растет, что может служить стимулом для профессионального развития и непрерывного обучения.

Спрос на специалистов в области DS продолжит расти, особенно на тех, кто владеет Python и SQL.
Новичкам стоит сосредоточиться на изучении Python и SQL, а также на понимании основ машинного обучения, чтобы улучшить свои шансы на трудоустройство.  
Знание Python и SQL может быть ключевым фактором для достижения высокой зарплаты в сфере DS.
Опытным специалистам следует продолжать развивать свои навыки в этих областях и следить за новыми технологиями и методами анализа данных.

# **Общий вывод по проекту**

**Предварительный анализ данных:** Рынок труда демонстрирует множество возможностей и разнообразие, что способствует конкуренции и развитию карьеры.  
**Детальный анализ вакансий:** Географическое распределение и информация о зарплате отражают тенденции рынка и важность гибкости условий труда.  
**Анализ работодателей:** Крупные компании и IT-сектор играют ключевую роль на рынке, требуя улучшения качества данных о вакансиях.  
**Предметный анализ:** Спрос на Data Science растет, особенно для специалистов с навыками Python и SQL, что влияет на зарплатные ожидания и карьерный рост.  


Анализ рынка труда показывает, что существует значительное количество вакансий, отражающее активность и разнообразие экономической среды. Крупные города и IT-отрасль демонстрируют наибольший спрос на специалистов, особенно в области Data Science, где важны навыки работы с Python и SQL. Рост зарплаты с увеличением опыта подчеркивает ценность профессионального развития.

Основной движущей силой является цифровизация экономики, которая увеличивает потребность в специалистах по обработке данных и аналитике.

Спрос на Data Science и смежные технологические навыки будет продолжать расти, поскольку данные становятся ключевым активом во всех отраслях. Работодателям следует инвестировать в обучение и развитие навыков своих сотрудников, а также предлагать конкурентоспособные условия для привлечения и удержания талантов. 